controllers/clustergroupupgrade_controller.go

```diff
@@ -973,34 +1141,148 @@ func (r *ClusterGroupUpgradeReconciler) reconcileResources(ctx context.Context,
 	return nil
 }
 
-func (r *ClusterGroupUpgradeReconciler) buildRemediationPlan(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) {
+func (r *ClusterGroupUpgradeReconciler) isClusterNonCompliantWithPolicy(clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, clusterName string, policy *unstructured.Unstructured) bool {
+	// Get the status part of the policy.
+	statusObject := policy.Object["status"].(map[string]interface{})
+	// Get the part of the status that holds the clusters' compliance.
+	statusCompliance := statusObject["status"]
+
+	// If all the clusters are compliant, return nil.
+	if statusCompliance == nil {
+		return false
+	}
+
+	subStatus := statusCompliance.([]interface{})
+
+	// Loop through all the NonCompliant clusters in the policy's status.
+	for _, crtSubStatusCrt := range subStatus {
+		crtSubStatusMap := crtSubStatusCrt.(map[string]interface{})
+		// If the cluster is NonCompliant, add it to the list of NonCompliant clusters.
+		if crtSubStatusMap["compliant"] == "NonCompliant" && clusterName == crtSubStatusMap["clustername"].(string) {
+			return true
+		}
+	}
+
+	return false
+}
+
+/* A policy's status has the format below. In this function we are creating a list with the clusternames
+   which are NonCompliant.
+
+status: (map[string]interface{})
+  compliant: NonCompliant
+  placement:
+  - placementBinding: operator-placementbinding
+    placementRule: operator-placementrules
+  status: ([]interface{})
+  - clustername: spoke3
+    clusternamespace: spoke3
+    compliant: NonCompliant
+  - clustername: spoke4
+    clusternamespace: spoke4
+    compliant: NonCompliant
+*/
+func (r *ClusterGroupUpgradeReconciler) getClustersNonCompliantWithPolicy(policy *unstructured.Unstructured) []string {
```

Comment:
I wonder if ACM has any golang libraries that we could leverage/use for some of these functions?

Reply:
Yes, it does. I actually tried playing with one on Friday, but ended up getting some errors and decided to better work on having something functional first and then look more into it. What I found was a library for placementRules.

Check this out: https://github.com/RamenDR/ramen/blob/aaab22aea138286a671046675f2526f405f25986/controllers/drplacementcontrol_controller.go#L977, https://github.com/RamenDR/ramen/blob/aaab22aea138286a671046675f2526f405f25986/controllers/drplacementcontrol_controller.go#L882
using this import: github.com/open-cluster-management/multicloud-operators-placementrule/pkg/apis/apps/v1

Unfortunately I didn't save the exact error, but it was complaining about the specific placementRule kind used by that package. I suspect an extra change was needed in SetupWithManager, but I chose to move forward with my stuff after spending some time and not being able to figure it out :disappointed: It's on my list of improvements though.

Reply:
Getting things functional is the first (most critical) step right now. We should keep this on our cleanup list and make use of the libraries where we can (after things are working)

Reply:
> 
@irinamihai When I try to use the ManagedCluster from open-cluster-management API, the issue I have probably is same with what you are having here.

The error I have is, 
ERROR   setup   problem running manager {"error": "no kind is registered for the type v1.ManagedCluster in scheme \"pkg/runtime/scheme.go:100\""}

I figured it out that you need to register the scheme for the APIs for mapping bewteens Kinds and their corresponding Go types.

Basically, in this example to use multicloud-operators-placementrule apps APIs,
in the main.go,
// import the pkg
import (
     appsv1 github.com/open-cluster-management/multicloud-operators-placementrule/pkg/apis/apps/v1
) 
// register the scheme
func init() {
  ...
  utilruntime.Must(appsv1.AddToScheme(scheme))
}

Then, the call appsv1.xxx should work.
Hope this is helpful.